File: src/mcts.cpp
#include "mcts.h"

#include <algorithm>
#include <cmath>

void apply_move(Board& board, int move, Stone color) {
    const int size = board.size();
    if (move == size * size) {
        board.pass();
    } else {
        board.play(move % size, move / size, color);
    }
}

float terminal_value(const Board& board, Stone to_play) {
    const float margin = board.score();
    if (margin == 0.0f) return 0.0f;  // jigo (integer komi)
    return (margin > 0) == (to_play == Stone::Black) ? 1.0f : -1.0f;
}

Search::Search(float c_puct, float dirichlet_alpha, float noise_fraction,
               std::uint64_t seed)
    : c_puct_(c_puct),
      dirichlet_alpha_(dirichlet_alpha),
      noise_fraction_(noise_fraction),
      rng_(seed) {}

std::vector<Node*> Search::descend(Node& root, Board& board,
                                   Stone& to_play,
                                   bool force_at_root) const {
    std::vector<Node*> path;
    path.push_back(&root);
    Node* node = &root;
    bool at_root = true;
    while (node->expanded()) {
        Edge* best = nullptr;
        if (at_root && force_at_root) {
            double best_deficit = 0.0;
            for (Edge& edge : node->edges) {
                const double forced =
                    kForcedPlayoutK *
                    std::sqrt(double(edge.prior) * double(node->visits));
                const double deficit = forced - edge.child->visits;
                if (deficit > best_deficit) {
                    best_deficit = deficit;
                    best = &edge;
                }
            }
        }
        if (!best) {
            const double sqrt_visits = std::sqrt(double(node->visits));
            double best_score = -1e30;
            for (Edge& edge : node->edges) {
                const double q = -edge.child->value();
                const double u = c_puct_ * edge.prior * sqrt_visits /
                                 (1 + edge.child->visits);
                if (q + u > best_score) {
                    best_score = q + u;
                    best = &edge;
                }
            }
        }
        at_root = false;
        apply_move(board, best->move, to_play);
        to_play = opponent(to_play);
        node = best->child.get();
        path.push_back(node);
    }
    return path;
}

void Search::expand(Node& node, const Board& board, Stone to_play,
                    const float* priors, bool allow_pass) const {
    const int points = board.size() * board.size();
    std::vector<int> legal = board.legal_moves(to_play);
    if (allow_pass || legal.empty()) {
        legal.push_back(points);
    }

    double total = 0.0;
    for (int move : legal) total += std::max(priors[move], 0.0f);

    node.edges.reserve(legal.size());
    for (int move : legal) {
        const float prior =
            total > 0 ? float(std::max(priors[move], 0.0f) / total)
                      : 1.0f / legal.size();
        node.edges.push_back({move, prior, std::make_unique<Node>()});
    }
}

void Search::backprop(const std::vector<Node*>& path, double value,
                      int visit_delta) {
    for (auto it = path.rbegin(); it != path.rend(); ++it) {
        (*it)->visits += visit_delta;
        (*it)->value_sum += value;
        value = -value;
    }
}

void Search::add_dirichlet_noise(Node& root) {
    std::gamma_distribution<double> gamma(dirichlet_alpha_, 1.0);
    std::vector<double> noise(root.edges.size());
    double total = 0.0;
    for (double& n : noise) {
        n = gamma(rng_);
        total += n;
    }
    if (total <= 0) return;
    for (std::size_t i = 0; i < root.edges.size(); i++) {
        root.edges[i].prior =
            (1 - noise_fraction_) * root.edges[i].prior +
            noise_fraction_ * float(noise[i] / total);
    }
}

std::vector<float> Search::policy(const Node& root, int points,
                                  float prune_forced_k) const {
    std::vector<float> pi(points + 1, 0.0f);
    const Edge* best = nullptr;
    for (const Edge& edge : root.edges) {
        if (!best || edge.child->visits > best->child->visits) {
            best = &edge;
        }
    }
    double total = 0.0;
    for (const Edge& edge : root.edges) {
        double visits = edge.child->visits;
        if (prune_forced_k > 0 && &edge != best) {
            visits -= prune_forced_k * std::sqrt(double(edge.prior) *
                                                 double(root.visits));
            if (visits < 1.0) visits = 0.0;
        }
        pi[edge.move] = float(visits);
        total += visits;
    }
    if (total > 0) {
        for (float& p : pi) p = float(p / total);
    }
    return pi;
}

int Search::select_move(const Node& root, int points, double temperature) {
    std::vector<float> pi = policy(root, points);
    if (temperature <= 0.0) {
        return int(std::max_element(pi.begin(), pi.end()) - pi.begin());
    }
    std::vector<double> weights(pi.size());
    for (std::size_t i = 0; i < pi.size(); i++) {
        weights[i] = pi[i] > 0 ? std::pow(double(pi[i]), 1.0 / temperature)
                               : 0.0;
    }
    std::discrete_distribution<int> dist(weights.begin(), weights.end());
    return dist(rng_);
}

File: train/mcts.py
"""PUCT Monte Carlo tree search (AlphaZero-style) over goboard positions.

The search is decoupled from the network: it takes an `evaluate`
callable mapping (board, to_play) to (priors, value), where priors is
an array of size*size + 1 move probabilities (last index = pass) and
value is the expected outcome in [-1, 1] for the side to play.

Values stored in a node are from the perspective of the player to move
at that node, so selection scores an edge with the negated child value.
"""

import math

import numpy as np

import goboard
from goboard import Board, Stone


class Node:
    __slots__ = ("prior", "visits", "value_sum", "children")

    def __init__(self, prior: float):
        self.prior = prior
        self.visits = 0
        self.value_sum = 0.0
        self.children: dict[int, Node] = {}

    def value(self) -> float:
        return self.value_sum / self.visits if self.visits else 0.0


def terminal_value(board: Board, to_play: Stone) -> float:
    """Outcome of a finished game for the side to play."""
    black_margin = board.score()
    if black_margin == 0.0:
        return 0.0  # jigo (integer komi)
    return 1.0 if (black_margin > 0) == (to_play == Stone.BLACK) else -1.0


class MCTS:
    def __init__(self, evaluate, c_puct: float = 1.5,
                 dirichlet_alpha: float = 0.3, noise_fraction: float = 0.25,
                 rng: np.random.Generator | None = None):
        self.evaluate = evaluate
        self.c_puct = c_puct
        self.dirichlet_alpha = dirichlet_alpha
        self.noise_fraction = noise_fraction
        self.rng = rng if rng is not None else np.random.default_rng()

    def run(self, board: Board, to_play: Stone, simulations: int,
            add_noise: bool = False) -> Node:
        root = Node(0.0)
        priors, value = self.evaluate(board, to_play)
        self.expand(root, board, to_play, priors, value)
        if add_noise:
            self.add_dirichlet_noise(root)
        for _ in range(simulations):
            self._simulate(root, board.copy(), to_play)
        return root

    def policy(self, root: Node, board_size: int) -> np.ndarray:
        """Visit-count distribution over size*size + 1 moves."""
        pi = np.zeros(board_size * board_size + 1, dtype=np.float32)
        for move, child in root.children.items():
            pi[move] = child.visits
        total = pi.sum()
        return pi / total if total > 0 else pi

    def select_move(self, root: Node, board_size: int,
                    temperature: float = 0.0) -> int:
        pi = self.policy(root, board_size)
        if temperature == 0.0:
            return int(pi.argmax())
        pi = pi ** (1.0 / temperature)
        pi /= pi.sum()
        return int(self.rng.choice(len(pi), p=pi))

    def _simulate(self, root: Node, board: Board, to_play: Stone) -> None:
        path, board, color = self.descend(root, board, to_play)
        if board.is_terminal():
            value = terminal_value(board, color)
        else:
            priors, value = self.evaluate(board, color)
            value = self.expand(path[-1], board, color, priors, value)
        self.backprop(path, value)

    def descend(self, root: Node, board: Board,
                to_play: Stone) -> tuple[list[Node], Board, Stone]:
        """Walk to a leaf, mutating `board` along the way."""
        node = root
        color = to_play
        path = [root]
        while node.children:
            move, node = self._select_child(node)
            apply_move(board, move, color)
            color = goboard.opponent(color)
            path.append(node)
        return path, board, color

    def backprop(self, path: list[Node], value: float,
                 visit_delta: int = 1) -> None:
        """`value` is from the perspective of the player to move at the
        leaf; the sign flips at every step back up. visit_delta=0 lets
        a virtual loss be corrected without recounting the visit."""
        for node in reversed(path):
            node.visits += visit_delta
            node.value_sum += value
            value = -value

    def _select_child(self, node: Node) -> tuple[int, "Node"]:
        sqrt_visits = math.sqrt(node.visits)
        best_score = -float("inf")
        best = None
        for move, child in node.children.items():
            q = -child.value()
            u = self.c_puct * child.prior * sqrt_visits / (1 + child.visits)
            if q + u > best_score:
                best_score = q + u
                best = (move, child)
        return best

    def expand(self, node: Node, board: Board, to_play: Stone,
               priors: np.ndarray, value: float) -> float:
        pass_move = board.size * board.size
        legal = board.legal_moves(to_play) + [pass_move]

        legal_priors = np.asarray([priors[m] for m in legal], dtype=np.float64)
        total = legal_priors.sum()
        if total > 0:
            legal_priors /= total
        else:
            legal_priors[:] = 1.0 / len(legal)
        for move, prior in zip(legal, legal_priors):
            node.children[move] = Node(float(prior))
        return float(value)

    def add_dirichlet_noise(self, root: Node) -> None:
        moves = list(root.children)
        noise = self.rng.dirichlet([self.dirichlet_alpha] * len(moves))
        for move, eta in zip(moves, noise):
            child = root.children[move]
            child.prior = ((1 - self.noise_fraction) * child.prior
                           + self.noise_fraction * eta)


def apply_move(board: Board, move: int, color: Stone) -> None:
    """Apply a flattened move index; size*size means pass."""
    size = board.size
    if move == size * size:
        board.pass_()
        return
    if not board.play(move % size, move // size, color):
        raise ValueError(f"illegal move {move} for {color}")


def uniform_evaluate(board: Board, to_play: Stone) -> tuple[np.ndarray, float]:
    """Network-free evaluator, useful for tests and sanity checks."""
    n = board.size * board.size + 1
    return np.full(n, 1.0 / n, dtype=np.float32), 0.0
